### PR TITLE
Pinecone fix vector search

### PR DESCRIPTION
**Pull request by henri123lemoine**
This PR:
- Adds many utils functions, most important of which are `get_embeddings` and co. `get_embeddings` calls the moderation endpoint before embedding whatever was *not* flagged, and returns the embeddings as well as the moderation results.
- Deals with PINECONE_METADATA_KEYS and removes `is_metadata_keys_equal` since it was doing nothing.
- Defines the session engine outside of make_session in order to reuse the same engine. Not sure this was necessary.
- Adds 'force_update' tag to pinecone_update so that the user can update a source in the pinecone db despite its mysql articles having pinecone_update_required be false.
- Adds finetuning. This includes 'finetuning_dataset.py' and 'training.py'. The former deals with the iterable dataset for the finetuning training, and the latter deals with the models and training process. This section has many remaining issues and needs to be improved substantially still. We should discuss issues with it at the meeting later today.
- Adds `query_vector`, `query_text`, and `get_embeddings_by_ids` methods to the PineconeDB class. The first two are just querying the pinecone db, and the third gets embeddings from (full-)ids. Used in the finetuning process.
- Adds the `pinecone_models.py` file, that includes the PineconeMetadata and PineconeEntry classes.
- Changed `update_pinecone.py` quite a bit. This includes adding chunks that got flagged to the article's comments. I don't know if something else about the article should be changed when certain chunks are flagged; to discuss in today's meeting.
- (Temporarily?) set the bias to 1 for all sources, so there's no favoring any source anymore. Biasing can be added later, but it's better to test it before just guessing a number and using it for the normal namespace.
- Adds pyproject.toml file for black to default to 100-char lines. Many of the cosmetic changes are linked to that, unfortunately. 

I realize it would have been better to split it into smaller branches.

### CODE REVIEW

File: align_data/pinecone/pinecone_models.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        "url",
+        "date_published",
+        "authors",
+        "text_chunks",
+        "embeddings",
+        pre=True,
+        always=True,
+    )
+    def empty_strings_not_allowed(cls, value):
+        if not str(value).strip():
```
Comment (mruwnik) [inline align_data/pinecone/pinecone_models.py:52]: can `value` be None? If so, you want `if value is None or not str(value).strip()` or something

**Review by mruwnik — COMMENTED** [review]
you might also want to add Black as a precommit hook

File: align_data/common/utils.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
+import logging
```
[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/common/utils.py:1]: yeah I agree, brother went to sleep but told me to check his code, improve it and fix it, so I can do that rn; utils related to embeddings / pinecone go to pinecone directory (unless we wanted a "embeddings" directory, which would be different from the pinecone directory, but idk)

File: align_data/common/utils.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+)
+
+hf_embeddings = (
+    HuggingFaceEmbeddings(
+        model_name=SENTENCE_TRANSFORMER_EMBEDDINGS_MODEL,
+        model_kwargs={"device": DEVICE},
+        encode_kwargs={"show_progress_bar": False},
+    )
+    if not USE_OPENAI_EMBEDDINGS
+    else None
```
[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/common/utils.py:53]: agree

File: align_data/common/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/common/utils.py:1]: I moved common/utils.py to pinecone/embedding_utils.py, but now I'm wondering, since it's used by finetuning/finetuning_dataset.py, wouldn't it be just in common/, but named something more descriptive like embedding_utils.py there? or merge finetuning/ and pinecone/ but they are quite separate

File: align_data/common/alignment_dataset.py [hunk 4]
```diff
@@ -249,9 +248,7 @@ def _load_outputted_items(self) -> Set[str]:
         with make_session() as session:
             return set(
                 session.scalars(
-                    select(Article.url)
-                    .join(Article.summaries)
-                    .filter(Summary.source == self.name)
+                    select(Article.url).join(Article.summaries).filter(Summary.source == self.name)
```
Comment (Thomas-Lemoine) [inline align_data/common/alignment_dataset.py:251]: isn't this just less readable?

File: align_data/db/session.py [hunk 5]
```diff
@@ -10,22 +10,24 @@
 
 logger = logging.getLogger(__name__)
 
+ENGINE = create_engine(DB_CONNECTION_URI, echo=False)
```
[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/db/session.py:13]: Is the difference between this and the previous implementation that, now, engine is created once during runtime instead of everytime make_session is called?

File: align_data/finetuning/finetuning_dataset.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+from typing import List, Tuple, Generator
+from collections import deque
+
+import torch
+from torch.utils.data import IterableDataset, get_worker_info
+from sqlalchemy.exc import OperationalError
+from sqlalchemy.sql import func
+
+from align_data.db.session import make_session
+from align_data.pinecone.embedding_utils import get_embedding
```
Comment (Thomas-Lemoine) [inline align_data/finetuning/finetuning_dataset.py:12]: seems like all these align_data.pinecone imports indicates that finetuning should be within pinecone, no? Or maybe both be within a bigger class, embedding? not sure though

File: align_data/finetuning/finetuning_dataset.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+class FinetuningDataset(IterableDataset):
+    def __init__(self, num_batches_per_epoch, cache_size=1280):
+        self.num_batches_per_epoch = num_batches_per_epoch
+        self.article_cache = deque(maxlen=cache_size)
+
+        self.text_splitter = ParagraphSentenceUnitTextSplitter()
+        self.pinecone_db = PineconeDB()
+
+        with make_session() as session:
+            self.all_article_ids = [row[0] for row in session.query(Article.id).all()]
```
[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/finetuning/finetuning_dataset.py:28]: I'm considering just making more streams similar to stream_pinecone_updates, like stream_all_article_ids, or wtv. would that be better, where article streams use db/session.py largely? ig the code for the streams get farther away from the code they are run in, which is a disadvantage

[the remark this answers is not in the document]
Comment (Thomas-Lemoine) [reply align_data/finetuning/finetuning_dataset.py:28]:
```py
def stream_all_article_ids(
        session: Session
):
    """Yield all article ids."""
    yield from (
        session.query(Article.id)
        .filter(Article.is_valid)
        .filter(or_(Article.confidence == None, Article.confidence > MIN_CONFIDENCE))
        .yield_per(1000)
    )
```
no idea if this works or is desirable

File: align_data/common/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: (done) added black to precommit hook
(done) utils moved and renamed

File: align_data/common/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: (done)

File: align_data/common/utils.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+            logger.warning(f"OpenAI Rate limit error. Trying again. Error: {e}")
+            raise
+        except APIError as e:
+            if "502" in str(e):
+                logger.warning(f"OpenAI 502 Bad Gateway error. Trying again. Error: {e}")
+            else:
+                logger.error(f"OpenAI API Error encountered: {e}")
+            raise
+        except (
+            InvalidRequestError,
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: (done)

File: align_data/common/utils.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+# --------------------
+# UTILITY FUNCTIONS
+# --------------------
+
+
+def get_recursive_type(obj):
+    """Get the nested type of a given object."""
+
+    if isinstance(obj, list):
+        return f"List[{get_recursive_type(obj[0]) if obj else 'Any'}]"
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: I just removed the `get_recursive_type` and `get_recursive_length` functions.

File: align_data/common/utils.py [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+
+# --------------------
+# MAIN FUNCTIONS
+# --------------------
+
+
+@handle_openai_errors
+def moderation_check(texts):
+    return openai.Moderation.create(input=texts)["results"]
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: yes

File: align_data/common/utils.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+            embeddings = hf_embeddings.embed_documents(texts)
+
+    # Bias adjustment
+    if bias := EMBEDDING_LENGTH_BIAS.get(source, 1.0):
+        embeddings = [[bias * e for e in embedding] for embedding in embeddings]
+
+    return embeddings
+
+
+def get_embeddings_or_none_if_flagged(
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: I'm not sure I understand how you want it to be split into two functions. TBH I'd delete it entirely, it's not being used anyway.

File: align_data/common/utils.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+    flagged_bools = [result["flagged"] for result in moderation_results]
+
+    non_flagged_texts = [text for text, flagged in zip(texts, flagged_bools) if not flagged]
+    non_flagged_embeddings = get_embeddings_without_moderation(
+        non_flagged_texts, source, engine, **kwargs
+    )
+
+    # Reconstruct the final list of embeddings with None for flagged texts
+    embeddings_iter = iter(non_flagged_embeddings)
+    embeddings = [None if flagged else next(embeddings_iter) for flagged in flagged_bools]
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: how about:
```py
embeddings = []
for flagged in flagged_bools:
    embeddings.append(None if flagged else non_flagged_embeddings.pop(0))
```
It's slightly slower because of the pop but time's really not the bottleneck anyway here.

File: align_data/common/utils.py [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+                    for j in range(len(obj))
+                ]
+            )
+        return lengths
+    return 1
+
+
+# --------------------
+# MAIN FUNCTIONS
+# --------------------
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/common/utils.py]: (done?) moved to `align_data/embeddings/embedding_utils.py`.

File: align_data/finetuning/finetuning_dataset.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/finetuning/finetuning_dataset.py:28]: Now using:
```py
def get_all_valid_article_ids(session: Session):
    """Return all valid article IDs."""
    query_result = (
        session.query(Article.id)
        .filter(Article.is_valid)
        .filter(or_(Article.confidence == None, Article.confidence > MIN_CONFIDENCE))
        .all()
    )
    return [item[0] for item in query_result]
```

File: align_data/finetuning/finetuning_dataset.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            if embedding is None:
+                embedding, _ = get_embedding(chunk)
+            embeddings.append(torch.tensor(embedding))
+
+        return embeddings
+
+    def _generate_pairs(
+        self, start=0, end=None, neg_pos_proportion=0.5
+    ) -> Generator[Tuple[List[float], List[float], int], None, None]:
+        end = end or self.total_articles
```
[the remark this answers is not in the document]
Comment (henri123lemoine) [reply align_data/finetuning/finetuning_dataset.py:77]: The advantage of creating a session in __iter__ is that the session is created only once per worker, which can save overhead compared to creating a new session for every batch generated in _generate_pairs.

File: align_data/pinecone/pinecone_models.py (hunk 1 above)
Comment (henri123lemoine) [reply align_data/pinecone/pinecone_models.py:52]: Depends if we want to add articles with no title/authors/url/date_published/text to the Pinecone DB. I'm fine either way.

File: align_data/finetuning/finetuning_dataset.py (hunk 6 above)
Comment (henri123lemoine) [reply align_data/finetuning/finetuning_dataset.py:12]: yeah did this

File: align_data/embeddings/pinecone/pinecone_models.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+                id=f"{self.hash_id}_{str(i).zfill(6)}",
+                values=self.embeddings[i],
+                metadata=PineconeMetadata(
+                    hash_id=self.hash_id,
+                    source=self.source,
+                    title=self.title,
+                    authors=self.authors,
+                    url=self.url,
+                    date_published=self.date_published,
+                    text=self.text_chunks[i],
```
Comment (mruwnik) [inline align_data/embeddings/pinecone/pinecone_models.py:72]: what about summaries? Will they also be embedded?

**Review by mruwnik — APPROVED** [review]
Looks good! At least the bits that I understood properly - I'll have to take your word for the models etc.

File: align_data/pinecone/pinecone_models.py (hunk 1 above)
Comment (Thomas-Lemoine) [reply align_data/pinecone/pinecone_models.py]: #154 this skips a pinecone entry if it has a falsy value for some field

File: align_data/embeddings/pinecone/pinecone_models.py (hunk 15 above)
Comment (Thomas-Lemoine) [reply align_data/embeddings/pinecone/pinecone_models.py:72]: good point, we haven't considered that yet

Comment (mruwnik) [reply align_data/embeddings/pinecone/pinecone_models.py:72]: It can be done in a later PR, though, if you want to merge this one

Comment (Thomas-Lemoine) [reply align_data/embeddings/pinecone/pinecone_models.py:72]: Although, from my understanding of it, if we did embed summaries, they would be their own PineconeEntry, so they would be added somewhere in PineconeUpdater's update(self, custom_sources: List[str], force_update: bool = False) function, right? Like, in my mind if we embed the summaries and add them as pinecone entries, they're treated like their own individual pinecone entry except for the fact that most of their fields have the same values as the article they are summarizing. If that's the case, I would imagine them created outside their PineconeEntry class.

If we wanted to have them be dealt automatically within PineconeEntry, one solution would be to add the article_summaries to get_text_chunks; this would mean each summary chunk would be seen as a chunk of the article, and embedded alongside it, but that seems a bit weak of a solution. Do you have other ideas?

Comment (Thomas-Lemoine) [reply align_data/embeddings/pinecone/pinecone_models.py:72]: Hmm, yeah I think it's worth merging this one and considering the embedding of summaries as a next step

**Review by Thomas-Lemoine — APPROVED** [review]

**Review by Thomas-Lemoine — APPROVED** [review]
